### PR TITLE
[BUGFIX][ENHANCEMENT] Robust Representation in Configuration of Both Legacy and New Datasource

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* Combining both the Legacy and the New Datasource configuration components in the same Datasource Configuration Schema and Datasource Configuration
* Utility methods for converting DataContextConfiguration to a yaml string and dictionary representations
* Updating DataContextV3 to use the new utility methods
* Removing duplicated method from DataContext

##### Notes
* Additional work will be needed to fully represent both the Legacy and the New Datasource configuration components in a way that provides Schema validation.  The present Pull Request is an important first step in Datasource unification.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [ENHANCEMENT], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
-
-
-

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

Thank you for submitting!

### CODE REVIEW

**Review by abegong — COMMENTED** [review]
I thought we were moving away from the pattern of typed configs for DataContext components (e.g. Datasources, Stores, Checkpoints, etc.)

@jcampbell , can you confirm when you get back?

In the meantime, let's hold off on merging this PR.

**Comment by alexsherstinsky** [comment]
> I thought we were moving away from the pattern of typed configs for DataContext components (e.g. Datasources, Stores, Checkpoints, etc.)
> 
> @jcampbell , can you confirm when you get back?
> 
> In the meantime, let's hold off on merging this PR.

@abegong @jcampbell -- from my understanding: we never planned to give up on Typed objects and Marshmallow for the components of DataContext Configuration.  The reason we suspended the Schema checking temporarily is that I have been having issues in getting it to work for deep DataConnector and ExecutionEngine class hierarchies with the reliance on flexible constructor parameters.  However, James and I are planning to meet this week to figure out the reason.  In the mean time, this PR fulfills on the idea of having both Legacy and New Datasource under the same configuration Schema and class (this approach came from the 1-1 James and I had yesterday).  I think that it should be merged (subject to any changes).  Thanks!

**Comment by abegong** [comment]
:+1:

For merging this PR, I'll defer to @jcampbell , since I wasn't on the call yesterday.

Different request: can we please create a design doc/style guide to capture these principles? Something along the lines of "How typing works in Great Expectations," published in Core Concepts.

I'd advocate for including an MVP version of the doc as part of this PR, but you could probably talk me into submitting it as a separate PR. :)

**Comment by alexsherstinsky** [comment]
> 👍
> 
> For merging this PR, I'll defer to @jcampbell , since I wasn't on the call yesterday.
> 
> Different request: can we please create a design doc/style guide to capture these principles? Something along the lines of "How typing works in Great Expectations," published in Core Concepts.
> 
> I'd advocate for including an MVP version of the doc as part of this PR, but you could probably talk me into submitting it as a separate PR. :)

@abegong -- the timing of meetings is such that I can add a draft design document in the present PR.  Thank you.

File: great_expectations/data_context/data_context_v3.py
Comment (jcampbell) [inline great_expectations/data_context/data_context_v3.py]: Can we just remove this entire method so we don't change the behavior of get_config?

Comment (jcampbell) [reply great_expectations/data_context/data_context_v3.py]: Or, should this become he base implementation?

File: great_expectations/data_context/types/base.py
Comment (jcampbell) [inline great_expectations/data_context/types/base.py]:
```suggestion
    def to_json_dict(self) -> dict:
```

Comment (jcampbell) [reply great_expectations/data_context/types/base.py]: Related: instead of inheriting from DictDot, could we inherit from SerializableDictDot now?

**Review by jcampbell — APPROVED** [review]
Looks good to me, but with a couple minor tweaks.

File: great_expectations/data_context/types/base.py
Comment (alexsherstinsky) [reply great_expectations/data_context/types/base.py]: This was done for DataContextConfig (committed).

File: great_expectations/data_context/data_context_v3.py
Comment (alexsherstinsky) [reply great_expectations/data_context/data_context_v3.py]: This was done (committed).

**Comment by alexsherstinsky** [comment]
@abegong During a meeting with @jcampbell we wanted to make the Schema document for Core Concepts as part of a separate PR.  The document that I shared is still valid and should be reviewed; it is just that the present PR is imminently mergeable.  Thank you.

**Pull request merged**
